feat: add Couchbase as Source and Tool

shyam-cb opened the pull request:
Added couchbase support to Genai Toolbox

anubhav756 commented on docs/en/resources/tools/couchbase-sql.md:
````diff
@@ -0,0 +1,64 @@
+## Example
+
+```yaml
+tools:
+  search_products_by_category:
+    kind: couchbase-sql
+    source: my-couchbase-instance
+    statement: |
+      SELECT p.name, p.price, p.description
+      FROM products p
+      WHERE p.category = $category AND p.price < $max_price
+      ORDER BY p.price DESC
+      LIMIT 10
+    description: |
+      Use this tool to get a list of products for a specific category under a maximum price.
+      Takes a category name, e.g. "Electronics" and a maximum price e.g 500 and returns a list of product names, prices, and descriptions.
+      Do NOT use this tool with invalid category names. Do NOT guess a category name, Do NOT guess a price.
+      Example:
+      {{
+          "category": "Electronics",
+          "max_price": 500
+      }}
+      Example:
+      {{
+          "category": "Furniture",
+          "max_price": 1000
+      }}
+    parameters:
+      - name: category
+        type: string
+        description: Product category name
+      - name: max_price
+        type: number
+        description: Maximum price (positive number)
````
nit: Can we make the indentation consistent?

anubhav756 commented on docs/en/resources/tools/couchbase-sql.md:
```diff
@@ -0,0 +1,64 @@
+        type: string
+        description: Product category name
+      - name: max_price
+        type: number
```
Shouldn't this be either `integer` or `float`?

shyam-cb replied:
changed to integer

shyam-cb commented on docs/en/resources/sources/couchbase.md:
```diff
@@ -0,0 +1,68 @@
+| client_key          | string   |    false     | Path to client key file for TLS authentication.                                                                             |
+| client_key_password | string   |    false     | Password for the client key.                                                                                                |
+| ca_cert             | string   |    false     | Path to CA certificate file.                                                                                                |
+| no_ssl_verify       | boolean  |    false     | If true, skip server certificate verification.                                                                              |
```
update the description

shyam-cb commented:
@anubhav756 How add secrets for integration testing ?

kurtisvg commented:
Hey @shyam-cb -- we'll have to figure out how to add Couchbase to our testing project and add some credentials as environment variables for the build pipeline. @duwenxin99 should be able to help with this.

shyam-cb commented:
> @duwenxin99 
Should I add Couchbase settings to integration.cloudbuild.yaml, and also where should I set the environment variables? I can use the substitutions section for bucket, scope, and collection name. Thanks.

duwenxin99 commented:
> > @duwenxin99
> > Should I add Couchbase settings to integration.cloudbuild.yaml, and also where should I set the environment variables? I can use the substitutions section for bucket, scope, and collection name. Thanks.

Hi @shyam-cb, I will update the cloudbuild file and set up a instance in our Toolbox testing project so you don't need to worry about testing for now. It would be great if you could test locally and verify that the implementations work correctly. Thanks!

shyam-cb commented:
> > > @duwenxin99
> > > Should I add Couchbase settings to integration.cloudbuild.yaml, and also where should I set the environment variables? I can use the substitutions section for bucket, scope, and collection name. Thanks.
> 
> Hi @shyam-cb, I will update the cloudbuild file and set up a instance in our Toolbox testing project so you don't need to worry about testing for now. It would be great if you could test locally and verify that the implementations work correctly. Thanks!

@duwenxin99 I have locally tested both unit and integration tests, as well as the workflow.

duwenxin99 commented:
> > > > @duwenxin99
> > > > Should I add Couchbase settings to integration.cloudbuild.yaml, and also where should I set the environment variables? I can use the substitutions section for bucket, scope, and collection name. Thanks.
> > 
> > 
> > Hi @shyam-cb, I will update the cloudbuild file and set up a instance in our Toolbox testing project so you don't need to worry about testing for now. It would be great if you could test locally and verify that the implementations work correctly. Thanks!
> 
> @duwenxin99 I have locally tested both unit and integration tests, as well as the workflow.

Hi @shyam-cb, we just got approved for using Couchbase on our testing project today. Could you grant me access to your forked repo so I could add testing configs to the PR? Thank you!

duwenxin99 commented:
Hi @shyam-cb, I have set up a Couchbase testing instance. A friendly ping to grant me access to your forked repo so I could update the test config. Thanks!

shyam-cb commented:
> Hi @shyam-cb, I have set up a Couchbase testing instance. A friendly ping to grant me access to your forked repo so I could update the test config. Thanks!

Hi @duwenxin99 I’ve just granted you access to the forked repository. Please feel free to add or update the configuration files as needed.

duwenxin99 commented:
/gcbrun

duwenxin99 commented:
> > Hi @shyam-cb, I have set up a Couchbase testing instance. A friendly ping to grant me access to your forked repo so I could update the test config. Thanks!
> 
> Hi @duwenxin99 I’ve just granted you access to the forked repository. Please feel free to add or update the configuration files as needed.

Thank you @shyam-cb, I have added the test configs and moved the helper functions around a bit. There are currently test failures for param tool and auth tool invocation calls. Could you fix them please? Thanks!

shyam-cb commented:
> @duwenxin99

Could you please share the test report? It’s working on my local machine.

duwenxin99 commented:
> > @duwenxin99
> 
> Could you please share the test report? It’s working on my local machine.

@shyam-cb 
The integration test `Invoke_my-auth-tool_with_auth_toke` is failing:
<img width="701" alt="Screenshot 2025-04-14 at 11 12 20 AM" src="https://github.com/user-attachments/assets/06c53a6e-cc5c-4295-a1f2-fcf262943888" />
<img width="822" alt="Screenshot 2025-04-14 at 11 12 30 AM" src="https://github.com/user-attachments/assets/675306cf-534f-4367-928a-2aa59a612f87" />
I ran it locally and it was failing for the same reason.

shyam-cb commented:
> @duwenxin99

SERVICE_ACCOUNT_EMAIL is required for the test case
TestCouchbaseToolEndpoints/Invoke_my-auth-tool_with_auth_token.

I have updated the integration.cloudbuild.yaml accordingly. However, the integration-test-pr is still failing. Could you please try again by pulling the latest changes from your local environment with SERVICE_ACCOUNT_EMAIL env set (same email used to set up the CLIENT_ID )?
Thanks

duwenxin99 commented:
> > @duwenxin99
> 
> SERVICE_ACCOUNT_EMAIL is required for the test case TestCouchbaseToolEndpoints/Invoke_my-auth-tool_with_auth_token.
> 
> I have updated the integration.cloudbuild.yaml accordingly. However, the integration-test-pr is still failing. Could you please try again by pulling the latest changes from your local environment with SERVICE_ACCOUNT_EMAIL env set (same email used to set up the CLIENT_ID )? Thanks

@shyam-cb I spent some time debugging and realized it was a firewall policy issue on the Cloud Build VM (which is why it works fine locally but not on the testing VM). I have set up another testing instance to bypass the firewall and now the integration test is passing. I will do some final cleanups and will merge this PR. Thank you very much for your contribution!

duwenxin99 commented:
/gcbrun

duwenxin99 commented:
> > > @duwenxin99
> > 
> > 
> > SERVICE_ACCOUNT_EMAIL is required for the test case TestCouchbaseToolEndpoints/Invoke_my-auth-tool_with_auth_token.
> > I have updated the integration.cloudbuild.yaml accordingly. However, the integration-test-pr is still failing. Could you please try again by pulling the latest changes from your local environment with SERVICE_ACCOUNT_EMAIL env set (same email used to set up the CLIENT_ID )? Thanks
> 
> @shyam-cb I spent some time debugging and realized it was a firewall policy issue on the Cloud Build VM (which is why it works fine locally but not on the testing VM). I have set up another testing instance to bypass the firewall and now the integration test is passing. I will do some final cleanups and will merge this PR. Thank you very much for your contribution!

nvm it's still failing. I'll let you know when I figure out the cause.

duwenxin99 commented:
Hi @shyam-cb, the integration test failure is due to data consistency issue. 
Originally, the test failed for a "null" result:
<img width="905" alt="Screenshot 2025-04-18 at 4 12 03 PM" src="https://github.com/user-attachments/assets/8d034d9a-1c7d-48bf-be11-63aeee999e0f" />
I added 1s timer after inserting test data and before running the test, the Tool invocation returned incomplete result: 
<img width="880" alt="Screenshot 2025-04-18 at 4 12 20 PM" src="https://github.com/user-attachments/assets/cccee096-8ce3-4a00-8615-6e44df09542b" />
After increasing the timer to 3s, the test now passes implying a complete result is returned.
However, I also saw that there is a parameter set to require request-level consistency [here](https://github.com/googleapis/genai-toolbox/blob/b632e3e4d1cd12033e6b2a1a71b2f8441ea22199/internal/tools/couchbase/couchbase.go#L105) so I assumed this implies a consistent query result.
Since the timer is for testing only and should be avoided in our code, I'd appreciate your expertise in implementing the proper fix to this issue. Thanks!

shyam-cb commented:
> duwenxin99

What Couchbase Server version are you currently using?
The previous code implementation, where tests passed successfully, used request_plus consistency and it is alway request_plus.
We’ve now modified the code to allow users to specify the query scan consistency in the source configuration.
I have not removed the timeout settings why the tests are failing

duwenxin99 commented:
> > duwenxin99
> 
> What Couchbase Server version are you currently using? Also cluster size. The previous code implementation, where tests passed successfully, used request_plus consistency and it is alway request_plus. We’ve now modified the code to allow users to specify the query scan consistency in the source configuration. I have not removed the timeout settings why integration tests are failing.

Hi @shyam-cb, the testing instance is using the version "7.6.4-5146-enterprise" and the cluster size is 1. The integration test is still passing (was failing previously for permission issue). What value to specify for the QueryScanConsistency field to make it strongly consistent? Could you also update the docs to include the new field?

shyam-cb commented:
> > > duwenxin99
> > 
> > 
> > What Couchbase Server version are you currently using? Also cluster size. The previous code implementation, where tests passed successfully, used request_plus consistency and it is alway request_plus. We’ve now modified the code to allow users to specify the query scan consistency in the source configuration. I have not removed the timeout settings why integration tests are failing.
> 
> Hi @shyam-cb, the testing instance is using the version "7.6.4-5146-enterprise" and the cluster size is 1. The integration test is still passing (was failing previously for permission issue). What value to specify for the QueryScanConsistency field to make it strongly consistent? Could you also update the docs to include the new field?

updated the docs

duwenxin99 commented:
> > > > duwenxin99
> > > 
> > > 
> > > What Couchbase Server version are you currently using? Also cluster size. The previous code implementation, where tests passed successfully, used request_plus consistency and it is alway request_plus. We’ve now modified the code to allow users to specify the query scan consistency in the source configuration. I have not removed the timeout settings why integration tests are failing.
> > 
> > 
> > Hi @shyam-cb, the testing instance is using the version "7.6.4-5146-enterprise" and the cluster size is 1. The integration test is still passing (was failing previously for permission issue). What value to specify for the QueryScanConsistency field to make it strongly consistent? Could you also update the docs to include the new field?
> 
> @duwenxin99 updated the docs

@shyam-cb Could you update the integration test to make sure it passes without the timer? Thanks!

shyam-cb commented:
@duwenxin99 what is the size of the cluster node ?

duwenxin99 commented:
Hi @shyam-cb, I did some rebase and cleanups, could you take the final look and let me know if you are ready to have it merged? Thanks!

shyam-cb commented:
Hi @duwenxin99 I have made only small change ie mad the queryScanConsistency to have default value zero instead of 2 can you rerun the integration testing again. If the integrations passes we can merge the PR.
Thanks

duwenxin99 commented:
> Hi @duwenxin99 I have made only small change ie mad the queryScanConsistency to have default value zero instead of 2 can you rerun the integration testing again. If the integrations passes we can merge the PR. Thanks

Hi @shyam-cb, the newly added primary index creation code is failing because the collection is not ready when we are trying to create the index. Could you also try and see if you have a `re-run failed checks` button on the top right corner when you click into the integration test? If you do, clicking it allows you to run the integration test. Thanks!
<img width="195" alt="Screenshot 2025-05-01 at 12 39 52 PM" src="https://github.com/user-attachments/assets/8f51d9da-da0e-42a3-b772-2775a2488fac" />

shyam-cb commented:
> > Hi @duwenxin99 I have made only small change ie mad the queryScanConsistency to have default value zero instead of 2 can you rerun the integration testing again. If the integrations passes we can merge the PR. Thanks
> 
> Hi @shyam-cb, the newly added primary index creation code is failing because the collection is not ready when we are trying to create the index. Could you fix it and also see if you have a `re-run failed checks` button on the top right corner when you click into the integration test? If you do, clicking it allows you to run the integration test. Thanks! <img alt="Screenshot 2025-05-01 at 12 39 52 PM" width="195" src="https://private-user-images.githubusercontent.com/117315983/439657683-8f51d9da-da0e-42a3-b772-2775a2488fac.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDYxMTk5ODUsIm5iZiI6MTc0NjExOTY4NSwicGF0aCI6Ii8xMTczMTU5ODMvNDM5NjU3NjgzLThmNTFkOWRhLWRhMGUtNDJhMy1iNzcyLTI3NzVhMjQ4OGZhYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUwMVQxNzE0NDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yNjM1YWMxNGJmNDc2M2JkNDNiMzNjMDYyOGU1ZTUyNGQzYzRkZTM2OTMwNjdlZjk1NGZmOTI0YmZiMDBkMjc3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.T6IQ0cwgsxi8thyg2Q5vWh6vL33cLBFdXkurkODQp7k">

@duwenxin99  I don't see the above options. can you run the integration test once again

duwenxin99 commented:
> > > Hi @duwenxin99 I have made only small change ie mad the queryScanConsistency to have default value zero instead of 2 can you rerun the integration testing again. If the integrations passes we can merge the PR. Thanks
> > 
> > 
> > Hi @shyam-cb, the newly added primary index creation code is failing because the collection is not ready when we are trying to create the index. Could you fix it and also see if you have a `re-run failed checks` button on the top right corner when you click into the integration test? If you do, clicking it allows you to run the integration test. Thanks! <img alt="Screenshot 2025-05-01 at 12 39 52 PM" width="195" src="https://private-user-images.githubusercontent.com/117315983/439657683-8f51d9da-da0e-42a3-b772-2775a2488fac.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDYxMTk5ODUsIm5iZiI6MTc0NjExOTY4NSwicGF0aCI6Ii8xMTczMTU5ODMvNDM5NjU3NjgzLThmNTFkOWRhLWRhMGUtNDJhMy1iNzcyLTI3NzVhMjQ4OGZhYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUwMVQxNzE0NDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yNjM1YWMxNGJmNDc2M2JkNDNiMzNjMDYyOGU1ZTUyNGQzYzRkZTM2OTMwNjdlZjk1NGZmOTI0YmZiMDBkMjc3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.T6IQ0cwgsxi8thyg2Q5vWh6vL33cLBFdXkurkODQp7k">
> 
> @duwenxin99 I don't see the above options. can you run the integration test once again

Is the timer still necessary with the retry logic?

duwenxin99 commented:
> > > > > Hi @duwenxin99 I have made only small change ie mad the queryScanConsistency to have default value zero instead of 2 can you rerun the integration testing again. If the integrations passes we can merge the PR. Thanks
> > > > 
> > > > 
> > > > Hi @shyam-cb, the newly added primary index creation code is failing because the collection is not ready when we are trying to create the index. Could you fix it and also see if you have a `re-run failed checks` button on the top right corner when you click into the integration test? If you do, clicking it allows you to run the integration test. Thanks! <img alt="Screenshot 2025-05-01 at 12 39 52 PM" width="195" src="https://private-user-images.githubusercontent.com/117315983/439657683-8f51d9da-da0e-42a3-b772-2775a2488fac.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDYxMTk5ODUsIm5iZiI6MTc0NjExOTY4NSwicGF0aCI6Ii8xMTczMTU5ODMvNDM5NjU3NjgzLThmNTFkOWRhLWRhMGUtNDJhMy1iNzcyLTI3NzVhMjQ4OGZhYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUwMVQxNzE0NDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yNjM1YWMxNGJmNDc2M2JkNDNiMzNjMDYyOGU1ZTUyNGQzYzRkZTM2OTMwNjdlZjk1NGZmOTI0YmZiMDBkMjc3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.T6IQ0cwgsxi8thyg2Q5vWh6vL33cLBFdXkurkODQp7k">
> > > 
> > > 
> > > @duwenxin99 I don't see the above options. can you run the integration test once again
> > 
> > 
> > Is the timer still necessary with the retry logic?
> 
> @duwenxin99 retry logic is just a fall back so the tests doesn't become flaky

Is there another way around? We want to avoid timers if we could.
